src/memory_manager.rs

```diff
@@ -630,6 +630,52 @@ pub fn is_mmtk_object(addr: Address) -> bool {
     SFT_MAP.get_checked(addr).is_mmtk_object(addr)
 }
 
+/// Find if there is an object with VO bit set for the given address range.
+/// This should be used instead of [`crate::memory_manager::is_mmtk_object`] for conservative stack scanning if
+/// the binding may have internal pointers on the stack.
+///
+/// The function cannot directly return an object reference. Instead, it returns
+/// an address range and guarantees the object ference is in the range.
+/// The reason behind this is that we use the VO
+/// bit to represent a valid object for every 8 bytes ([`crate::util::is_mmtk_object::VO_BIT_REGION_SIZE`]).
+/// We use VO bits to find the object for an internal pointer.
+/// When we find a set VO bit, we only know that in the 8 bytes there is an object, and we cannot know where
+/// exactly the object is. The binding needs to use their knowledge about the alignment
+/// and offset for object references to find out the object reference from the return value.
+/// See Case 2 in [`crate::memory_manager::is_mmtk_object`] for more explanation.
+///
+/// Note that, we only consider pointers that points to an address that is equal or after the in-object addresss
+/// (i.e. the address returned from [`crate::vm::ObjectModel::ref_to_address`]), and within the allocation
+/// as 'internal pointers'. To be precise, for each object ref `obj_ref`, internal pointers are in the range
+/// `[ObjectModel::ref_to_address(obj_ref), ObjectModel::ref_to_object_start(obj_ref) + ObjectModel::get_current_size(obj_ref))`.
+/// If a binding defines internal pointers differently, calling this method is undefined behavior.
+/// If this is the case for you, please submit an issue or engage us on Zulip to discuss more.
+///
+/// Note that, in the similar situation as [`crate::memory_manager::is_mmtk_object`], the binding should filter
+/// out obvious non-pointers (e.g. alignment check, bound check, etc) before calling this function to avoid unnecessary
+/// cost. This method is not cheap.
+///
+/// To minimize the cost, the user should also use a small `max_search_bytes`.
+///
+/// Argument:
+/// * `internal_ptr`: The internal pointer to start. We search backwards from this internal pointer to find the base reference.
+/// * `max_search_bytes`: The maximum number of bytes we may search for an object with VO bit set.
+#[cfg(feature = "is_mmtk_object")]
+pub fn find_object_from_internal_pointer<VM: VMBinding>(
```

Comment:
I suggest moving this method to the module `mmtk::util::is_mmtk_object`.  As a Rust package, it is better to use modules to categorize functions instead of putting all static methods in one magical class like JikesRVM.

We should move `memory_manager::is_mmtk_object` into the `mmtk:util::is_mmtk_object` module, too, and leave a stub here for existing bindings that depend on `memory_manager::is_mmtk_object`.  The doc comments on `memory_manager::is_mmtk_object` needs to be updated, too, because we now require `ObjectReference` to be word-aligned, and we can remove some of the confusing descriptions there.  `memory_manager::is_mmtk_object` should return `Option<ObjectReference>`, too, for the sake of symmetry, and also because it is safe to convert an `Address` to `ObjectReference` once we know VO bit is set for its corresponding in-object address.  We may update `is_mmtk_object` in https://github.com/mmtk/mmtk-core/pull/1159 because that PR changes the requirement of `ObjectReference`.

Reply:
Sounds good.

Reply:
I changed the return type for `is_mmtk_object`, and updated the document. But I did not move the functions to the `vo_bit` module. The current control flow is `memory_manager` -> SFT -> policy -> `vo_bit`. If we move the functions to `vo_bit`, the control flow would be `memory_manager` -> `vo_bit` -> SFT -> policy -> `vo_bit`, which seems confusing.

Reply:
There is a module named `mmtk::util::is_mmtk_object`.  It currently only holds the constant `VO_BIT_REGION_SIZE` which the VM binding needs to know the granularity of the VO bits.  We can move the implementation of the `is_mmtk_object` function there.  But I acknowledge that the module name may not be the best (`mmtk::util::is_mmtk_object::is_mmtk_object` looks confusing, too).  If we don't have a better plan for reorganizing the functions, we can leave the `is_mmtk_object` function in the `memory_manager` module for now.

Reply:
I moved the actual implementation of the two methods to `is_mmtk_object`. But those are not public, and the public methods are still in `memory_manager`.